formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/conversion/StmtConversionVisitor.kt

```diff
@@ -281,18 +283,40 @@ object StmtConversionVisitor : FirVisitor<ExpEmbedding, StmtConversionContext>()
             }
         }
 
+    private fun List<FirExpression>.withPureVarargsHandled(
+        data: StmtConversionContext
+    ) =
+        flatMap { arg ->
+            when (arg) {
+                is FirVarargArgumentsExpression -> {
+                    arg.arguments.map(data::convert)
+                }
+
+                else -> listOf(data.convert(arg))
+            }
+        }
+
     override fun visitFunctionCall(functionCall: FirFunctionCall, data: StmtConversionContext): ExpEmbedding {
         val symbol = functionCall.toResolvedCallableSymbol() as? FirFunctionSymbol<*>
             ?: throw NotImplementedError("Only functions are expected as callables of function calls, got ${functionCall.toResolvedCallableSymbol()}")
 
         when (val forAllLambda = functionCall.extractFormverFirBlock { isInvariantBuilderFunctionNamed("forAll") }) {
             null -> {
-                val callee = data.embedFunction(symbol)
-                return callee.insertCall(
-                    functionCall.functionCallArguments.withVarargsHandled(data, callee),
-                    data,
-                    data.embedType(functionCall.resolvedType),
-                )
+                if (data.isPureFunction(symbol)) {
```

Comment:
I think this separation of responsibilities is not ideal.  The StmtConversionVisitor should not care whether the function it embeds is pure or not; it just needs to get a callable object from the ProgramConverter, and then put a call in it.  Deciding how to embed that function shouldn't be up to it.